pkg/controller/lambda/function/setup.go

```diff
@@ -0,0 +1,499 @@
+	case string(svcapitypes.State_Inactive):
+		cr.SetConditions(xpv1.Unavailable())
+	}
+	return obs, nil
+}
+
+func postCreate(_ context.Context, cr *svcapitypes.Function, resp *svcsdk.FunctionConfiguration, cre managed.ExternalCreation, err error) (managed.ExternalCreation, error) {
+	if err != nil {
+		return managed.ExternalCreation{}, err
+	}
+	meta.SetExternalName(cr, aws.StringValue(resp.FunctionName))
+	cre.ExternalNameAssigned = true
+	return cre, nil
+}
+
+func preDelete(_ context.Context, cr *svcapitypes.Function, obj *svcsdk.DeleteFunctionInput) (bool, error) {
+	obj.FunctionName = aws.String(meta.GetExternalName(cr))
+	return false, nil
+}
+
+// nolint:gocyclo
+func isUpToDate(cr *svcapitypes.Function, obj *svcsdk.GetFunctionOutput) (bool, error) {
+
+	// Compare CODE
+	if !isUpToDateCode(cr, obj) {
+		return false, nil
+	}
+
+	// Compare CONFIGURATION
+	if strings.Compare(aws.StringValue(cr.Spec.ForProvider.Description), aws.StringValue(obj.Configuration.Description)) != 0 {
+		return false, nil
+	}
+
+	if !isUpToDateEnvironment(cr, obj) {
+		return false, nil
+	}
+
+	// Connection settings for an Amazon EFS file system.
+	if !isUpToDateFileSystemConfigs(cr, obj) {
+		return false, nil
+	}
+
+	if strings.Compare(aws.StringValue(cr.Spec.ForProvider.Handler), aws.StringValue(obj.Configuration.Handler)) != 0 {
+		return false, nil
+	}
+
+	if strings.Compare(aws.StringValue(cr.Spec.ForProvider.KMSKeyARN), aws.StringValue(obj.Configuration.KMSKeyArn)) != 0 {
+		return false, nil
+	}
+
+	// The function's layers (https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html).
+	if !isUpToDateLayers(cr, obj) {
+		return false, nil
+	}
+
+	// set default
+	if aws.Int64Value(cr.Spec.ForProvider.MemorySize) != 0 &&
+		aws.Int64Value(cr.Spec.ForProvider.MemorySize) != aws.Int64Value(obj.Configuration.MemorySize) {
+		return false, nil
+	}
+
+	if strings.Compare(aws.StringValue(cr.Spec.ForProvider.Role), aws.StringValue(obj.Configuration.Role)) != 0 {
+		return false, nil
+	}
+
+	if strings.Compare(aws.StringValue(cr.Spec.ForProvider.Runtime), aws.StringValue(obj.Configuration.Runtime)) != 0 {
+		return false, nil
+	}
+	// Set default
+	if aws.Int64Value(cr.Spec.ForProvider.Timeout) != 0 &&
+		aws.Int64Value(cr.Spec.ForProvider.Timeout) != aws.Int64Value(obj.Configuration.Timeout) {
+		return false, nil
+	}
+
+	if !isUpToDateTracingConfig(cr, obj) {
+		return false, nil
+	}
+
+	if !isUpToDateSecurityGroupIDs(cr, obj) {
+		return false, nil
+	}
+
+	addTags, removeTags := diffTags(cr.Spec.ForProvider.Tags, obj.Tags)
+	return len(addTags) == 0 && len(removeTags) == 0, nil
+
+}
+
+// isUpToDateEnvironment checks if FunctionConfiguration EnvironmentResponse Variables are up to date
+func isUpToDateEnvironment(cr *svcapitypes.Function, obj *svcsdk.GetFunctionOutput) bool {
+	// Handle nil pointer refs
+	envVars := map[string]*string{}
+	awsVars := map[string]*string{}
+	if !reflect.ValueOf(cr.Spec.ForProvider.Environment).IsZero() &&
```

Comment:
Can we check for length equality and then iterate over one of the maps and return in case there is any mismatch with the other? `reflect` incurs some performance cost and unpredictability in the code in general.

Reply:
yes - it has been updated for better comparison